qiskit/circuit/random/utils.py

```diff
@@ -118,6 +86,256 @@ def random_circuit(
+        applying a 2Q gate on the selected qubit-pair.
+        prob_conditional (float): Probability less than 1.0, this is used to control the
+        occurence of conditionals in the circuit.
+
+    Returns:
+        QuantumCircuit: constructed circuit
+
+    Raises:
+        CircuitError: when invalid options given
+        ValueError: when the probabilities for qubit-pairs are passed and they do not add to 1.0
+        and when a negative probability exists.
+    """
+    pydi_graph, pydi_graph_node_map, _, _ = interaction_graph
+    edges_probs = pydi_graph.edges()
+
+    # Just a switch for the probability weighted selection of qubit-pairs.
+    # If any value of the probability is None or the whole probability list is None,
+    # then, the probability weighted selection of qubit-pair would be turned off.
+    prob_weighted_mapping = not None in edges_probs
+
+    if prob_weighted_mapping:
+        for prob in edges_probs:
+            if prob < 0:
+                raise ValueError("The probability should be positive")
+
+    if prob_weighted_mapping and not np.isclose(np.sum(edges_probs), 1.000, rtol=0.001):
+        raise ValueError(
+            "The sum of all the probabilities of a qubit-pair to be selected is not 1.0"
+        )
+
+    if pydi_graph_node_map is not None:
+        qubits = list(pydi_graph_node_map.keys())
+    else:
+        n_q = pydi_graph.num_nodes()
+        qubits = [Qubit(QuantumRegister(n_q, "q"), idx) for idx in pydi_graph.nodes()]
+    num_qubits = len(qubits)
+
+    if num_qubits == 0:
+        return QuantumCircuit()
+
+    if max_operands < 1 or max_operands > 2:
+        raise CircuitError("max_operands must be between 1 and 2")
+
+    max_operands = max_operands if num_qubits > max_operands else num_qubits
+
+    edge_list = list(pydi_graph.edge_list())
+    num_edges = len(edge_list)
+
+    if num_edges == 0 or max_operands == 1:
+        return random_circuit(
+            num_qubits=num_qubits,
+            depth=depth,
+            max_operands=1,
+            measure=measure,
+            conditional=conditional,
+            reset=reset,
+            seed=seed,
+        )
+
+    gates_1q, gates_2q, _, _ = _get_gates()
+
+    if reset:
+        gates_1q.append((Reset, 1, 0))
+
+    gates_2q = np.array(
+        gates_2q, dtype=[("class", object), ("num_qubits", np.int64), ("num_params", np.int64)]
+    )
+    gates_1q = np.array(gates_1q, dtype=gates_2q.dtype)
+
+    qc = QuantumCircuit(qubits)
+
+    if measure or conditional:
+        cr = ClassicalRegister(num_qubits, "c")
+        qc.add_register(cr)
+
+    if seed is None:
+        seed = np.random.randint(0, np.iinfo(np.int32).max)
+    rng = np.random.default_rng(seed)
+
+    qubits = np.array(qc.qubits, dtype=object, copy=True)
+
+    stop = False
+    layer_idx = 0
+    edges_used = {edge: 0 for edge in edge_list}
+
+    while not stop:
+
+        gate_2q_choice = rng.choice(gates_2q, size=1)[0]
+        control, target = tuple(
```

Comment:
normalized edge weights represent the probability with which each qubit-pair interaction is inserted into a layer.

Reply:
This is a huge performance drop!!, yet to find a way to calculate the total number of edges/2Q gates that would be required in the entire circuit. ( and, thus do the rng.choice once ).
since, in the choices we need to have each qubit-pair at-least 'depth' number of times and the selection is a random process, so, any, heuristic approach would be error prone :)